Fix source order sorting.

reitermarkus opened the pull request:
Fixes https://github.com/rust-lang/rust-bindgen/issues/2556.

Previously, headers included with `-include` were not added to the `includes` map since they don't have a source file. This is now fixed.

Also, some wrong match arms were removed/fixed.

reitermarkus committed:
Fix test on macOS.

reitermarkus committed:
Gather all includes, even ones without a source file.

reitermarkus commented:
cc @glandium @emilio

reitermarkus commented on bindgen-tests/tests/tests.rs:
```diff
@@ -468,9 +468,9 @@ fn test_mixed_header_and_header_contents() {
-        .header(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/headers/char.h"))
         .header_contents("test.h", "int bar(const char* a);")
         .header_contents("test2.h", "float bar2(const char* b);")
+        .header(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/headers/char.h"))
```
Note: This doesn't change anything, it simply reflects the order these are stored internally and eventually generated in.

emilio commented on bindgen/clang.rs:
```diff
@@ -555,36 +543,41 @@ impl Cursor {
+            (Some((file2, offset2)), Some((other_file2, other_offset2)))
+                if file2 == other_file2 =>
+            {
+                offset2.cmp(&other_offset2)
```
I don't think this is enough. You need to somehow deal with transitive includes don't you?

reitermarkus replied:
Can you give an example for what you think might get sorted wrong? Will try to add a test for this then.

Note that when recursing, the recursive `visit_sorted` gets sorted in and of itself, so this may actually work without dealing with transitive includes explicitly.

emilio replied:
I don't have a concrete test-case because it would depend on the concrete sorting algorithm, but not dealing with recursive includes would mean that you break the `Ord` invariants, as in, you might have a, b, and c, such that `a < b`, `b < c`, but `a == c` or so, breaking https://doc.rust-lang.org/std/cmp/trait.Ord.html#corollaries, and potentially invariants of the sorting algorithm. Need to take a look at the new approach.

pvdrz replied:
I agree, we shouldn't rely on the fact that sort works for the testcases we have so far even though we don't have an actual ordering as it is not transitive. 

I'd be happy to fix this in an upcoming PR if that's ok with both of you.

emilio commented on bindgen/clang.rs:
```diff
@@ -555,36 +543,41 @@ impl Cursor {
             }
-            (None, None) => cmp::Ordering::Equal,
+            // Otherwise, keep the original sorting.
+            _ => cmp::Ordering::Equal,
```
I guess this gets the fallback, but is this ok? Seems this could end up with weird cases where the partial ord invariants are not held up and we end up with really weird sorting right?

reitermarkus replied:
Remove the fallback, all cases should be handled explicitly now.

pvdrz replied:
This might be a bit evil but what happens if you do the following:
```h
// a.h
#ifndef A_H
#define A_H

struct foo {};

#include "b.h"

struct baz {
    struct bar field;
};

#endif

```
```h
//b.h
#include "a.h"

struct bar {
  struct foo field;
};
``
Does this ordering guarantee that we don't loop forever because `a.h` is included in `b.h` which includes `a.h`?

reitermarkus replied:
@pvdrz, yes, we won't loop forever, but there is still an error here.

If `a.h` is the main header file and we compare `baz` with `bar`, we check the include location of `baz`, which is erroneously inside `b.h`, resulting in `bar` wrongly being sorted after `baz`.

The fix for this is easy though: Simply ignore include directives which include the main header.

pvdrz replied:
:+1: could you add this as a test?

reitermarkus committed:
Add test for nested includes.

reitermarkus committed:
Handle all cases in `cmp_by_source_order`.

bors-servo commented:
:umbrella: The latest upstream changes (presumably ee980e1afeedc2afadafa6d8bdface54b5027e2d) made this pull request unmergeable. Please resolve the merge conflicts.

reitermarkus committed:
Add source order test for headers that are siblings.

reitermarkus commented on bindgen/clang.rs:
```diff
@@ -555,36 +543,154 @@ impl Cursor {
+                    // compare the include locations of `b.h` and `c.h` in `a.h` instead.
+                    include_location = ctx.included_file_location(file2);
+                    other_include_location =
+                        ctx.included_file_location(other_file2);
```
@emilio, I added another test which actually reaches this case.

reitermarkus committed:
Make `SourceLocation` an `enum`.

reitermarkus commented:
I refactored `SourceLocation` to be an `enum` and the sorting is now split between `SourceLocation` and `IncludeLocation`. In my opinion it is a bit easier to follow that way.

reitermarkus committed:
Don't count include locations of main header file.

reitermarkus commented:
@pvdrz, I fixed the edge case you gave and added a test for it.

reitermarkus commented:
I changed all file names from `String` to `Path` since we need to canonicalize them before comparing them, e.g. running your example with `cargo run -- a.h` would actually try to compare the strings `"a.h"` and `"./a.h"` before.

`ParseCallbacks::include_file` should probably also be changed to take a `&Path` instead of `&str`. Also its documentation

```rust
/// This will be called on every file inclusion, with the full path of the included file.
```

was actually wrong before (depending on whether you define “full path” as “absolute path”), since it would actually also pass relative paths.

emilio reviewed (commented):
I still think this is quite complicated for the benefit it brings fwiw... :/

reitermarkus commented:
@emilio, do you have an alternative? The status quo is simply  incorrect in some cases.

The only alternative I can see is to never generate constants for variable-like macros and only ever generate Rust macros instead. That way there is never a naming or ordering conflict.

Note that this is also needed for the same reason for function-like macros and functions in https://github.com/rust-lang/rust-bindgen/pull/2369.

reitermarkus commented:
No, it only broke some cases due to the sorting being wrong/incomplete previously. Otherwise nothing should break, except for some auto-generated names having a different index.

alex commented:
Confirmed that this works for one of my projects.

reitermarkus commented:
Thanks for testing, @alex.

@emilio, @pvdrz, can you have another look at this?

reitermarkus commented:
@pvdrz, @emilio, please have another look here, thank you.

reitermarkus commented:
@pvdrz, @emilio, another ping here.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 5ff913ab3349db1905cfbbf44a9121cbc223249d) made this pull request unmergeable. Please resolve the merge conflicts.

reitermarkus commented:
@pvdrz, @emilio, can you have another look at this?

emilio reviewed (commented):
I still think we should have some measurements of the overhead of the source order sorting, and turn it off by default if it's non-trivial.

Some of the refactoring like refactoring to use the clang::SourceLocation enum seems nice. Though that said, I'd rather keep the CXFile rather than reading the path and doing IO to absolutize it eagerly for most callers that don't need that...

reitermarkus commented:
> turn it off by default

This is a prerequisite for supporting complex macros (https://github.com/rust-lang/rust-bindgen/pull/2369), so it can't be optional.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 46c06e5ef11c01680edd9ca250ed0c32b6cdc9e9) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 285eb56cf5a02ee778880b8e1b7bb2e8118a5a88) made this pull request unmergeable. Please resolve the merge conflicts.

reitermarkus committed:
Don't eagerly absolutize source file paths.

reitermarkus commented:
@pvdrz, @emilio please have another look. 

> I'd rather keep the `CXFile` rather than reading the path and doing IO to absolutize it eagerly for most callers that don't need that...

I have re-added the `File` struct as `SourceFile` now to avoid eagerly absolutizing every path.

reitermarkus committed:
Use name instead of absolute path for checking allow/blocklist.

reitermarkus committed:
Use name instead of absolute path for diagnostics.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 3b5ce9c5861cd2e97e9789f5b686238656abd2d6) made this pull request unmergeable. Please resolve the merge conflicts.

reitermarkus commented:
@pvdrz, @emilio please have another look here.

bors-servo commented:
:umbrella: The latest upstream changes (presumably 93648e43b79c0382594aa351fbc219534bc24d41) made this pull request unmergeable. Please resolve the merge conflicts.
